SilverSolver/assignment2/binary_tree.py

```diff
@@ -0,0 +1,61 @@
+class BinaryNode:
+
+    def __init__(self, value, parent=None, child1=None, child2=None):
+        self.value  = value
+        self.parent = parent
+        self.child1 = child1
+        self.child2 = child2
+
+    def get_ancestors(self, prev_ancestors=None, value=True):
+        if not prev_ancestors:
+            prev_ancestors = []
+        if not self.parent:
+            return prev_ancestors
+        else:
+            if value:
+                prev_ancestors.append(self.parent.value)
+            else:
+                prev_ancestors.append(self.parent)
+            return self.parent.get_ancestors(prev_ancestors=prev_ancestors, value=value)
+
+    def get_lowest_common_ancestor(self, another_node, value=True):
+        self_anc = self.get_ancestors(value=False)
+        another_anc = another_node.get_ancestors(value=False)
+        low_c_anc = None
+        while(self_anc and another_anc):
+            if self_anc[-1] == another_anc[-1]:
+                low_c_anc = self_anc[-1]
+                self_anc = self_anc[:-1]
+                another_anc = another_anc[:-1]
+            else:
+                if value and low_c_anc:
+                    return low_c_anc.value
+                else:
+                    return low_c_anc
+        if value and low_c_anc:
+            return low_c_anc.value
+        else:
+            return low_c_anc
+
+class BinaryTree:
```

Comment:
What are  the advantages of using two classes? I found it simpler to use only a class Node with a recursive constructor. In any case, there is no important differences between the two approach.

Reply:
I supposed that class BinaryTree will be used to construct tree itself, and BinaryNode will contain information about the particular node. So they have different meaning, in my opinion and that's why can be divided. If you are sure that it is bad, I can re-design this part of the program in one of the following ways:
a) make one function "from_iterable" with the same meaning as constructor of the class BinaryTree
b) turn the constructior of the class BinaryTree into some method of BinaryNode
But I'm not sure whether it is worth the time or not.

Reply:
A designer knows he has achieved perfection not when there is nothing left to add, but when there is nothing left to take away.
  -- Antoine de Saint-Exupery 

Both solutions works, no need to rewrite, it would be a waste of time, I just want to share a few things, I feel the BInaryTree constructor is not needed as the same feature can be implemented in the Node constructor, depending on whether the optional left and right are provided

***Recursive data structure**
Here is conceptual approach as to why I prefer one class. With BinaryNode and BinaryTree, we get the zoom level locked respectively to the minimum granularity and the maximum granularity. With one class that represent a subtree, we have one view that generalizes both approach. The one class approach is akin to fractals where each node can be zoomed in and present the same structure, and same methods are available.

What if the BinaryTree gets so big that it does not fit in memory anymore? Then you would split it in several subtrees and host each subtree on a different memory server. Would you make each subtree a BinaryNode and keep them in another container? Would you create a third class? I personally find the one class approach more flexible.

**Inappropriate intimacy**
Two objects needs to be compatible, as they use  methods or attributes from each other, this compatiblity needs to be understood and respected for team members wanting to expand this code.
This pattern of Inappropriate intimacy is descrbied here with other software programming wisdom:
https://blog.codinghorror.com/code-smells/
(feel free to read and disagree)

**code is a liability, not a an asset**
Less objects usually means shorter code, low cognitive overhead, less bugs. Less code is also less effort for the team or the reviewers. Code always grow so complex that the velocity in adding new features decrease as the effort to understand the code and the fear of breaking it because we do not understand 100% of it. Here are the kind of features that would need to be implemented where I would add more objects: logging, profiling/statistics, access and authorization checks, parallelism, storage, memory redundancy, performance optimization.
Just to share some of my experience with OOP: I have worked on two projects where the inheritance chain would go more than 10 levels high, it is slow to understand what to modify when adding a feature.

Reply:
Thanks for detailed explanation. I didn't think about this data structure in terms of scalability. If we consider BinaryNode as a subtree, than BinaryTree is really excessive. These objects are really have too close meaning.